rust/src/mqtt/logger.rs

```diff
@@ -0,0 +1,326 @@
+                js.set_uint("qos", msg.header.qos_level as u64)?;
+                js.set_bool("retain", msg.header.retain)?;
+                js.set_bool("dup", msg.header.dup_flag)?;
+                js.set_uint("message_id", unsuback.message_id as u64)?;
+                if let Some(codes) = &unsuback.reason_codes {
+                    js.open_array("reason_codes")?;
+                    for t in codes {
+                        js.append_uint(*t as u64)?;
+                    }
+                    js.close()?; // reason_codes
+                }
+                js.close()?; // unsuback
+            }
+            MQTTOperation::PINGREQ => {
+                js.open_object("pingreq")?;
+                js.set_uint("qos", msg.header.qos_level as u64)?;
+                js.set_bool("retain", msg.header.retain)?;
+                js.set_bool("dup", msg.header.dup_flag)?;
+                js.close()?; // pingreq
+            },
+            MQTTOperation::PINGRESP => {
+                js.open_object("pingresp")?;
+                js.set_uint("qos", msg.header.qos_level as u64)?;
+                js.set_bool("retain", msg.header.retain)?;
+                js.set_bool("dup", msg.header.dup_flag)?;
+                js.close()?; // pingresp
+            },
+            MQTTOperation::AUTH(ref auth) => {
+                js.open_object("auth")?;
+                js.set_uint("qos", msg.header.qos_level as u64)?;
+                js.set_bool("retain", msg.header.retain)?;
+                js.set_bool("dup", msg.header.dup_flag)?;
+                js.set_uint("reason_code", auth.reason_code as u64)?;
+                if let Some(properties) = &auth.properties {
+                    js.open_object("properties")?;
+                    for prop in properties {
+                        prop.set_json(js)?;
+                    }
+                    js.close()?; // properties
+                }
+                js.close()?; // auth
+            },
+            MQTTOperation::DISCONNECT(ref disco) => {
+                js.open_object("disconnect")?;
+                js.set_uint("qos", msg.header.qos_level as u64)?;
+                js.set_bool("retain", msg.header.retain)?;
+                js.set_bool("dup", msg.header.dup_flag)?;
+                if let Some(reason_code) = &disco.reason_code {
+                    js.set_uint("reason_code", *reason_code as u64)?;
+                }
+                if let Some(properties) = &disco.properties {
+                    js.open_object("properties")?;
+                    for prop in properties {
+                        prop.set_json(js)?;
+                    }
+                    js.close()?; // properties
+                }
+                js.close()?; // disconnect
+            },
+            MQTTOperation::TRUNCATED(ref trunc) => {
+                js.open_object(&message_code_string(trunc.original_message_type))?;
+                js.set_uint("qos", msg.header.qos_level as u64)?;
```

Comment:
There is duplicated code for all 3 `header` fields, is not there ?

Reply:
Do you mean the `qos`, `retain` and `dup` ones? True, but I didn't find a way to handle those before branching out into the different message types because with the new stateful JSONBuilder the calls to `js.set_*()` need to come after the opening of the type-specific sub-object (e.g. `js.open_object("connect")?;`) which obviously needs to be in the `match` branches because before we dont know what type we have. I don't see that as much of a problem here than just a minor style issue.
I can put them in a separate inlined function and call that in each branch to reduce code duplication if that is enough to address this concern.

Reply:
This is indeed a minor style issue

Inline function seems a good solution to me.

Another solution may be in a generic `js.open_object("header")?;` before the `match`
What do you think about it ?
Does it make the use of json output simpler ? to find every log with a certain sos for instance, whatever the message type ? (does that make sense ?)